src/sources/http.rs

```diff
@@ -0,0 +1,339 @@
+    fn build(
+        &self,
+        _: &str,
+        _: &GlobalOptions,
+        out: mpsc::Sender<Event>,
+    ) -> crate::Result<super::Source> {
+        let source = SimpleHttpSource {
+            encoding: self.encoding.clone(),
+            headers: self.headers.clone(),
+        };
+        source.run(self.address, "".to_string(), out)
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn source_type(&self) -> &'static str {
+        "http"
+    }
+}
+
+fn add_headers(
+    mut events: Vec<Event>,
+    headers_config: &[String],
+    headers: HeaderMap,
+) -> Vec<Event> {
+    for header_name in headers_config {
+        let value = headers.get(header_name).map(HeaderValue::as_bytes).unwrap_or(b"");
+        for event in events.iter_mut() {
+            event.as_mut_log().insert(header_name as &str, value);
+        }
+    }
+
+    events
+}
+
+fn decode_body(body: impl Buf, enc: Encoding) -> Result<Vec<Event>, ErrorMessage> {
+    let mut rdr = BufReader::new(body.reader())
+        .lines()
+        .filter_map(|res| {
+            res.map_err(|error| error!(message = "Error reading request body", ?error))
+                .ok()
+        })
+        .filter(|s| !s.is_empty());
+
+    match enc {
+        Encoding::Text => Ok(rdr.map(Event::from).collect()),
+        Encoding::Ndjson => Ok(rdr
+            .map(|j| json_parse_object(json_parse(j)?))
+            .collect::<Result<_, _>>()?),
+        Encoding::Json => {
+            let first_line = rdr.next();
+            let second_line = rdr.next();
+            if first_line.is_some() && second_line.is_none() {
+                //need exactly one line
+                json_parse_array_of_object(json_parse(first_line.unwrap())?)
+            } else {
+                Err(json_error())
+            }
+        }
+    }
+}
+
+fn json_parse(json_str: String) -> Result<JsonValue, ErrorMessage> {
+    serde_json::from_str(&json_str).map_err(|_| json_error())
+}
+
+fn json_parse_object(value: JsonValue) -> Result<Event, ErrorMessage> {
+    let mut event = Event::new_empty_log();
+    let log = event.as_mut_log();
+    log.insert(event::TIMESTAMP.clone(), Utc::now()); // Add timestamp
+    match value {
+        JsonValue::Object(map) => {
+            flatten(log, map);
+            Ok(event)
+        }
+        _ => Err(json_error()),
+    }
+}
+
+fn json_parse_array_of_object(value: JsonValue) -> Result<Vec<Event>, ErrorMessage> {
+    match value {
+        JsonValue::Array(v) => {
+            v.into_iter()
+                .map(json_parse_object)
+                .collect::<Result<_,_>>()
+        }
+        JsonValue::Object(map) => {
+            //treat like an array of one object
+            Ok(vec![json_parse_object(JsonValue::Object(map))?])
+        }
+        _ => Err(json_error()),
+    }
+}
+
+fn json_error() -> ErrorMessage {
+    ErrorMessage::new(StatusCode::BAD_REQUEST, "Bad JSON".to_string())
```

Comment:
Maybe this should support some error context? Eg "Expected Array or Object, got Number"